mam/v1/receiver.go

```diff
@@ -0,0 +1,138 @@
+package mam
+
+import (
+	"github.com/iotaledger/iota.go/api"
+	"github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/converter"
+	"github.com/iotaledger/iota.go/transaction"
+	"github.com/iotaledger/iota.go/trinary"
+)
+
+// Receiver implements a receiver for MAM-Messages.
+type Receiver struct {
+	api     API
+	mode    ChannelMode
+	sideKey trinary.Trytes
+}
+
+// NewReceiver returns a new receiver.
+func NewReceiver(api API) *Receiver {
+	return &Receiver{
+		api:     api,
+		mode:    ChannelModePublic,
+		sideKey: consts.NullHashTrytes,
+	}
+}
+
+// SetMode sets the channel mode.
+func (r *Receiver) SetMode(m ChannelMode, sideKey trinary.Trytes) error {
+	if m != ChannelModePublic && m != ChannelModePrivate && m != ChannelModeRestricted {
+		return ErrUnknownChannelMode
+	}
+	if m == ChannelModeRestricted {
+		if sideKey == "" {
+			return ErrNoSideKey
+		}
+		r.sideKey = sideKey
+	}
+	r.mode = m
+	return nil
+}
+
+// Mode returns the channel mode.
+func (r *Receiver) Mode() ChannelMode {
+	return r.mode
+}
+
+// SideKey returns the channel's side key.
+func (r *Receiver) SideKey() trinary.Trytes {
+	return r.sideKey
+}
+
+// Receive tries to receive all messages from the specified root and returns them along with the next root.
+func (r *Receiver) Receive(root trinary.Trytes) (trinary.Trytes, []string, error) {
+	rootTrits, err := trinary.TrytesToTrits(root)
+	if err != nil {
+		return "", nil, err
+	}
+
+	address, err := makeAddress(r.mode, rootTrits, r.sideKey)
+	if err != nil {
+		return "", nil, err
+	}
+
+	txs, err := r.api.FindTransactionObjects(api.FindTransactionsQuery{Addresses: trinary.Hashes{address}})
+	if err != nil {
+		return "", nil, err
+	}
+
+	bundles := map[trinary.Trytes][]transaction.Transaction{}
+	for _, tx := range txs {
+		if v, ok := bundles[tx.Bundle]; ok {
+			bundles[tx.Bundle] = append(v, tx)
+		} else {
+			bundles[tx.Bundle] = []transaction.Transaction{tx}
+		}
+	}
+
+	var nextRoot trinary.Trytes
+	messages := []string{}
+	for _, bundle := range bundles {
+		candidateTxs := []transaction.Transaction{}
+		for _, tx := range bundle {
+			if tx.CurrentIndex == 0 {
+				candidateTxs = append(candidateTxs, tx)
```

Comment:
If we found the tail transaction of a given bundle, we can break out of this for loop immediately. Please add a break statement. If there were multiple txs with `currentIndex` = 0 in the same bundle, then something would be wrong anyway.

Reply:
Ok, that makes sense. If I add this break, there will be a maximum of one `candidateTx` which makes two loops obsolete.